zeus/monitor/carbon.py

```diff
@@ -54,7 +62,7 @@ def get_current_carbon_intensity(self) -> float:
         pass
 
     @abc.abstractmethod
-    def get_recent_carbon_intensity(self) -> dict[str, float]:
+    def get_recent_carbon_intensity(self) -> dict[datetime, float]:
```

Comment:
Question: When I get a dictionary that maps 2024 December 18th 10:33 PM ET as the key to a certain carbon intensity number, what does that mean?  Carbon intensity of the hour that includes  2024 December 18th 10:33 PM ET, i.e., 2024 December 18th 10:00 - 10:59 PM ET? The *duration* and *boundary* of the timeframe that corresponds to each carbon intensity value is unclear from this API contract.

Reply:
I believe in that example, the duration is from 2024 December 18th 10:00 - 10:59 PM ET.

If I make a fetch for live carbon intensity at 2024 December 18th 10:33 PM ET, the datetime recorded in the response is 2024 December 18th 10:00 ET. This will continue to be the live carbon intensity until the current datetime becomes 2024 December 18th 11:00 ET.

There shouldn't be a situation where 2024 December 18th 10:33 PM ET is a key in the map, the dictionary only includes keys of "whole hours" (2024 December 18th 10:00, 2024 December 18th 11:00, 2024 December 18th 12:00, etc.)

Reply:
Okay, so for ElectricityMaps, a datetime with 10 PM maps to the carbon intensity value of 10 PM to 11 PM. And we're doing all this back and forth because this is ambiguous in the API contract. What if someone implemented a new carbon intensity client that return 11 PM to mean the carbon intensity value for 10 PM to 11 PM? Or, what if another service has a granularity of 30 minutes, and returns dictionaries with keys 11:00 PM, 11:30 PM, 12:00 AM, etc.?

Reply:
Hmmm, ok this makes sense. I think what we can do is have each specific carbon intensity class store an attribute with the polling frequency and we can pass that to polling process. As for time ranges, would most simple be having the carbon intensity providers map the beginning of each window with the carbon intensity?

For example, if the granularity of updates for the carbon intensity provider is 30 minutes, carbon intensity provider will provider:
{10:00PM -> 300, 10:30PM->345, 11:00PM->343}, where 300 is the carbon intensit for 10:00-10:30PM.

Reply:
Yeah, I agree that making "the datetime key must mark the beginning of the period" part of the contract is the way to go. And for the granularity, I can think of something like this:

```python
@property
@abstractmethod
def update_period(self) -> int:
    """How long each carbon intensity value in the history dict remains current, in seconds."""
    return 3600
```

Or, `datetime.timedelta(hours=1)` should also work.

Reply:
Similar question-- should we also set up a similar method that can return how often you need to fetch from the API? ElectricityMaps needs to be fetched every 23 hours because thats how far back they give recent carbon intensity values for but other APIs might be different?

Reply:
Perhaps that can be inferred from calling `len` on the return value of `get_recent_carbon_intensity`.

Reply:
I ended up creating a similar `@property` method. I was thinking that this could save 1 API request and allows for the programmer to have finer control over the polling frequency from APIs.